From ed0a92d206ea96af0c8cd5b1406f8a6091285cc9 Mon Sep 17 00:00:00 2001
From: wwidjaja <w.widjaja@intel.com>
Date: Thu, 20 Feb 2020 18:35:52 +0800
Subject: [PATCH] PONRTSYS-4533 - when bootcore loaded to start of FMT1 try to
 auto set the DDR region

  - set the protection DDR region automatically based on size & FMT1
    base
---
 arch/mips/lantiq/grx500_bootcore/ngi_prx300.c | 51 +++++++++++++++++++++++++++
 1 file changed, 51 insertions(+)

diff --git a/arch/mips/lantiq/grx500_bootcore/ngi_prx300.c b/arch/mips/lantiq/grx500_bootcore/ngi_prx300.c
index 11c1715a8b81..59d2f278e2bb 100644
--- a/arch/mips/lantiq/grx500_bootcore/ngi_prx300.c
+++ b/arch/mips/lantiq/grx500_bootcore/ngi_prx300.c
@@ -27,6 +27,52 @@
 
 #define SSX_REGSIZE 0x00100000
 
+#if CONFIG_BOOTCORE_LOAD_ADDR == 0xffffffff88000000
+/* enable only if bootcore linux aligned with FMT1 start */
+#define AUTO_DDR_REGION_PROTECT
+#endif /* CONFIG_BOOTCORE_LOAD_ADDR == 0xffffffff88000000 */
+
+#define VAL_NGI_DDR_REGION_SIZE_16MB  0x79
+#define VAL_NGI_DDR_REGION_SIZE_32MB  0x81
+#define VAL_NGI_DDR_REGION_SIZE_64MB  0x89
+#define VAL_NGI_DDR_REGION_SIZE_128MB 0x91
+
+#ifdef AUTO_DDR_REGION_PROTECT
+
+#if defined(CONFIG_TOS_SIZE_16M)
+#define VAL_NGI_DDR_REGION_SIZE VAL_NGI_DDR_REGION_SIZE_16MB
+#elif defined(CONFIG_TOS_SIZE_32M)
+#define VAL_NGI_DDR_REGION_SIZE VAL_NGI_DDR_REGION_SIZE_32MB
+#elif defined(CONFIG_TOS_SIZE_64M)
+#define VAL_NGI_DDR_REGION_SIZE VAL_NGI_DDR_REGION_SIZE_64MB
+#elif defined(CONFIG_TOS_SIZE_128M)
+#define VAL_NGI_DDR_REGION_SIZE VAL_NGI_DDR_REGION_SIZE_128MB
+#else
+#define VAL_NGI_DDR_REGION_SIZE 0
+#endif
+
+#define SUB_4KEC_REG_BASE 0x14900000u
+#define SUB_4KEC_REG_SIZE 0x200
+#define AHB_CPU_PAGE_1_OFFSET 0x1D4
+
+static u32 get_base_ddr_fmt1(void)
+{
+	u32 base_fmt1 = 0;
+	void __iomem* sub_4kec_base = 0;
+
+	sub_4kec_base = ioremap_nocache( SUB_4KEC_REG_BASE, SUB_4KEC_REG_SIZE);
+	if (NULL == sub_4kec_base) {
+		pr_err("prx300-ngi: Error ioremap sub 4kec!");
+		return base_fmt1;
+	}
+
+	base_fmt1 = (__raw_readl(sub_4kec_base+AHB_CPU_PAGE_1_OFFSET) & 0x0000FFFF) << 16;
+
+	return base_fmt1;
+}
+
+#endif /* AUTO_DDR_REGION_PROTECT */
+
 static u32 permission_wr(void __iomem* base, u32 offset, u32 val)
 {
 	__raw_writel(val, (volatile void __iomem *)(base+offset));
@@ -87,9 +133,14 @@ void set_permission(void)
 			REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_0, ALL_CPUS);
 	permission_wr(ssx1_base,
 			REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_0, ALL_CPUS);
+	#ifdef AUTO_DDR_REGION_PROTECT
+	 permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_ADDR_MATCH_1, (get_base_ddr_fmt1()|VAL_NGI_DDR_REGION_SIZE));
+	#else /* AUTO_DDR_REGION_PROTECT */
 	/* protect DDR region 0x3E000000~0x40000000 32MB */
 	permission_wr(ssx1_base,
 			REG0_TDDR_PM_TDDR_PM_ADDR_MATCH_1, 0x3E000081);
+	#endif /* AUTO_DDR_REGION_PROTECT */
 	permission_wr(ssx1_base,
 			REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_1, M4KEC_ONLY);
 	permission_wr(ssx1_base,
